pkg/discover/discover.go

```diff
@@ -0,0 +1,179 @@
+)
+
+// Structs for handling the information that comes from the `app.yaml`
+type App struct {
+	Argocd_app_namespace              string
+	Argocd_app_source_repo_url        string
+	Argocd_app_source_target_revision string
+	Argocd_app_source_path            string
+	Application_folder                string
+	CreateNamespace                   bool   `yaml:"createNamespace"`
+	Name                              string `yaml:"name"`
+	Namespace                         string `yaml:"namespace"`
+	Project                           string `yaml:"project"`
+	Source                            Source `yaml:"source"`
+}
+
+type Source struct {
+	Helm_repo     string `yaml:"helm_repo"`
+	Chart_name    string `yaml:"chart_name"`
+	Chart_version string `yaml:"chart_version"`
+}
+
+// Unmarshals a .yaml file into the app struct
+func (app *App) GetValuesFromYamlFile(path string) error {
+	file, err := os.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	err = yaml.Unmarshal(file, &app)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//
+func Discover(folder string) error {
+	logging.Debug("folder: %v\n", folder)
+
+	yamlFiles, err := GetFiles(folder)
+	if err != nil {
+		return err
+	}
+
+	for _, path := range yamlFiles {
+		logging.Debug("path: %s\n", path)
+
+		var app App
+
+		app.GetValuesFromYamlFile(path)
+		app.getArgoCDEnvs()
+
+		// Remove the base from the path
+		tmpString := strings.ReplaceAll(path, folder+"/", "")
+		tmpStringArray := strings.Split(tmpString, "/")
+
+		// We can then assume, that the last two entris in the array represents the `app.yaml` file and the application folder
+		// this gives us the app path without having to assume, that the name in the `app.yaml` file and the name of the folder it's in is not the same
+		tmpString = strings.Join(tmpStringArray[:len(tmpStringArray)-1], "/")
+
+		app.Application_folder = tmpString
+
+		logging.Debug("App: %v\n", app)
+
+		app.GenArgoCDApp()
+	}
+
+	return nil
+}
+
+// Using the Golang template engine to generate the ArgoCD app
+func (app *App) GenArgoCDApp() error {
+	template, err := template.New("tempalte").Parse(getTempalte())
+	if err != nil {
+		return err
+	}
+
+	err = template.Execute(os.Stdout, app)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Trying to fetch ArgoCD environments
+func (app *App) getArgoCDEnvs() {
+	app.Argocd_app_namespace = os.Getenv("ARGOCD_APP_NAMESPACE")
+	app.Argocd_app_source_path = os.Getenv("ARGOCD_APP_SOURCE_PATH")
+	app.Argocd_app_source_repo_url = os.Getenv("ARGOCD_APP_SOURCE_REPO_URL")
+	app.Argocd_app_source_target_revision = os.Getenv("ARGOCD_APP_SOURCE_TARGET_REVISION")
+}
+
+// Looks up all files in the root, and checks if it contains the 'argocd-discover' apiVersion
+func GetFiles(folder string) ([]string, error) {
+	yamlFiles := []string{}
+	err := filepath.Walk(folder, func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if strings.Contains(path, ".yaml") {
```

Comment:
You don't want to match `foo.yaml.backup` or something weird.
```suggestion
		if strings.HasSuffix(path, ".yaml") {
```

Reply:
Good point